core/tracepoint/enable_spec.rb

```diff
@@ -99,4 +99,386 @@ def test; end
+    describe 'target: option' do
+      before :each do
+        ScratchPad.record []
+      end
+
+      it 'enables trace point for specific location' do
+        trace = TracePoint.new(:call) do |tp|
+          ScratchPad << tp.method_id
+        end
+
+        obj = Object.new
+        def obj.foo; end
+        def obj.bar; end
+
+        trace.enable(target: obj.method(:foo)) do
+          obj.foo
+          obj.bar
+        end
+
+        ScratchPad.recorded.should == [:foo]
+      end
+
+      it 'traces all the events triggered in specified location' do
+        trace = TracePoint.new(:line, :call, :return, :b_call, :b_return) do |tp|
+          ScratchPad << tp.event
+        end
+
+        obj = Object.new
+        def obj.foo
+          bar
+          -> {}.call
+        end
+        def obj.bar; end
+
+        trace.enable(target: obj.method(:foo)) do
+          obj.foo
+        end
+
+        ScratchPad.recorded.uniq.sort.should == [:call, :return, :b_call, :b_return, :line].sort
+      end
+
+      it 'does not trace events in nested locations' do
+        trace = TracePoint.new(:call) do |tp|
+          ScratchPad << tp.method_id
+        end
+
+        obj = Object.new
+        def obj.foo
+          bar
+        end
+        def obj.bar
+          baz
+        end
+        def obj.baz
+        end
+
+        trace.enable(target: obj.method(:foo)) do
+          obj.foo
+        end
+
+        ScratchPad.recorded.should == [:foo]
+      end
+
+      describe 'option value' do
+        it 'excepts Method' do
+          trace = TracePoint.new(:call) do |tp|
+            ScratchPad << tp.method_id
+          end
+
+          obj = Object.new
+          def obj.foo; end
+
+          trace.enable(target: obj.method(:foo)) do
+            obj.foo
+          end
+
+          ScratchPad.recorded.should == [:foo]
+        end
+
+        it 'excepts UnboundMethod' do
+          trace = TracePoint.new(:call) do |tp|
+            ScratchPad << tp.method_id
+          end
+
+          klass = Class.new do
+            def foo; end
+          end
+
+          unbound_method = klass.instance_method(:foo)
+          trace.enable(target: unbound_method) do
+            klass.new.foo
+          end
+
+          ScratchPad.recorded.should == [:foo]
+        end
+
+        it 'excepts Proc' do
+          trace = TracePoint.new(:b_call) do |tp|
+            ScratchPad << tp.lineno
+          end
+
+          block = proc {}
+          _, lineno = block.source_location
+
+          trace.enable(target: block) do
+            block.call
+          end
+
+          ScratchPad.recorded.should == [lineno]
+          lineno.should be_kind_of(Integer)
+        end
+
+        it 'excepts RubyVM::InstructionSequence' do
+          trace = TracePoint.new(:call) do |tp|
+            ScratchPad << tp.method_id
+          end
+
+          obj = Object.new
+          def obj.foo; end
+
+          iseq = RubyVM::InstructionSequence.of(obj.method(:foo))
```

Comment:
`RubyVM` is a MRI-specific class, and so should not be part of Ruby specs, but only MRI tests/specs.
i.e., I think MRI would not want other implementations to define `RubyVM`.

Reply:
@eregon  So should I remove these `RubyVM::InstructionSequence`-related specs?

Reply:
I'm interested in this tracepoint. How can it be implemented on other interpreter implementations?

Reply:
TracePoint itself is public and portable Ruby API, and I don't see issues with `TracePoint.enable(target:)`.

But `RubyVM` and everything under is MRI-specific AFAIK.

I'm not too happy about `TracePoint#instruction_sequence`, because that obviously cannot be supported on other Ruby implementations: https://bugs.ruby-lang.org/issues/15287#note-10

> @eregon So should I remove these `RubyVM::InstructionSequence`-related specs?

I'll give it some thoughts first, but at least it wouldn't make sense to count RubyVM specs as errors/failures on other implementations.

Reply:
@eregon What would be the equivalent on JRuby/TruffleRuby?

Reply:
> @eregon What would be the equivalent on JRuby/TruffleRuby?

@ioquatix For TruffleRuby, there is no eqauivalent to `RubyVM::InstructionSequence` presently (but TracePoint is generally supported).
The portable interface to Ruby VMs for Ruby code is Ruby code (text).
Exposing the bytecode or other intermediate representations prevent changing it (or requires careful versioning), which can prevent optimizations, and exposing it allows the user to break the VM too easily (it's a security issue and it's hard to fully verify bytecode or other intermediate representations).

Reply:
> TracePoint itself is public and portable Ruby API, and I don't see issues with `TracePoint.enable(target:)`.

In other words, I think it's easy to implement `TracePoint.enable(target:)` on TruffleRuby when the target is a Method, UnboundMethod or Proc, but not for a `RubyVM::InstructionSequence` target.

Reply:
@eregon I thought about this issue too, and I thought some kind of abstract `class Source` would make sense. https://bugs.ruby-lang.org/issues/6012

Reply:
@andrykonchin I removed the specs referencing `RubyVM`, as that can only be supported by MRI, see https://github.com/ruby/spec/issues/662.